Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift

```diff
@@ -9,201 +9,1069 @@
+                )
+
+            case .merging(let task, var buffer, let upstreamContinuations, let upstreamsFinished, .none):
+                // There is not outstanding downstream continuation so we must buffer the element
+                // This happens if we race our upstream sequences to produce elements
+                // and the _losers_ are signalling their produced element
+                state = .modifying
+
+                buffer.append(element)
+
+                state = .merging(
+                    task: task,
+                    buffer: buffer,
+                    upstreamContinuations: upstreamContinuations,
+                    upstreamsFinished: upstreamsFinished,
+                    downstreamContinuation: nil
+                )
+
+                return .none
+
+            case .upstreamThrew:
+                // Another upstream already produced an error so we just drop the new element
+                return .none
+
+            case .finished:
+                // Since cancellation is cooperative it might be that child tasks
+                // are still producing elements after we finished.
+                // We are just going to drop them since there is nothing we can do
+                return .none
+
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        /// Actions returned by `upstreamFinished()`.
+        enum UpstreamFinishedAction {
+            /// Indicates that the task and the upstream continuations should be cancelled.
+            case cancelTaskAndUpstreamContinuations(
+                task: Task<Void, Never>,
+                upstreamContinuations: [CheckedContinuation<Void, Error>]
+            )
+            /// Indicates that the downstream continuation should be resumed with `nil` and
+            /// the task and the upstream continuations should be cancelled.
+            case resumeContinuationWithNilAndCancelTaskAndUpstreamContinuations(
+                downstreamContinuation: CheckedContinuation<Element?, Error>,
+                task: Task<Void, Never>,
+                upstreamContinuations: [CheckedContinuation<Void, Error>]
+            )
+            /// Indicates that nothing should be done.
+            case none
+        }
+
+        mutating func upstreamFinished() -> UpstreamFinishedAction {
+            switch state {
+            case .initial:
+                preconditionFailure("Invalid state")
+
+            case .merging(let task, let buffer, let upstreamContinuations, var upstreamsFinished, let .some(downstreamContinuation)):
+                // One of the upstreams finished
+                precondition(buffer.isEmpty, "We are holding a continuation so the buffer must be empty")
+
+                // First we increment our counter of finished upstreams
+                upstreamsFinished += 1
+
+                // We should change the 2 when we get variadic generics
```

Comment:
might be worth it to abstract this to a static inline variable which ought to have the same perf as this. but then give it a name.

Reply:
Agreed, I also wanna change this around since I want to be able to support the merge2 and merge3 algorithm with the same state machine.